fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/metadata/RecordMetaDataBuilderTest.java

```diff
@@ -370,4 +420,103 @@ public void updateRecords() {
+        assertEquals(3L, metaData.getIndex("MySimpleRecord$num_value_3_indexed").getSubspaceKey());
+
+        // Test setting the counter without enabling it
+        e = assertThrows(MetaDataException.class, () -> RecordMetaData.newBuilder()
+                .setRecords(TestRecords1Proto.getDescriptor())
+                .setSubspaceKeysCounter(4L));
+        assertEquals("Counter-based subspace keys not enabled", e.getMessage());
+
+        // Test setting the counter to a value not greater than the current value
+        e = assertThrows(MetaDataException.class, () -> RecordMetaData.newBuilder()
+                .enableCounterBasedSubspaceKeys()
+                .setRecords(TestRecords1Proto.getDescriptor())
+                .setSubspaceKeysCounter(3L));
+        assertEquals("Subspace keys counter must be set to a value greater than its current value (3)", e.getMessage());
+
+        // Set to a random number
+        long randomCounter = ThreadLocalRandom.current().nextLong(1, Long.MAX_VALUE - 10);
+        builder = RecordMetaData.newBuilder().enableCounterBasedSubspaceKeys().setSubspaceKeysCounter(randomCounter).setRecords(TestRecords1Proto.getDescriptor());
+        metaData = builder.build(true);
+        assertNotNull(metaData.getIndex("MySimpleRecord$str_value_indexed"));
+        assertNotNull(metaData.getIndex("MySimpleRecord$num_value_unique"));
+        assertNotNull(metaData.getIndex("MySimpleRecord$num_value_3_indexed"));
+        assertEquals(randomCounter + 1, metaData.getIndex("MySimpleRecord$str_value_indexed").getSubspaceKey());
+        assertEquals(randomCounter + 2, metaData.getIndex("MySimpleRecord$num_value_unique").getSubspaceKey());
+        assertEquals(randomCounter + 3, metaData.getIndex("MySimpleRecord$num_value_3_indexed").getSubspaceKey());
+
+        // enable and set counter in the proto.
+        RecordMetaDataProto.MetaData.Builder protoBuilder = RecordMetaDataProto.MetaData.newBuilder()
+                .setRecords(TestRecords1Proto.getDescriptor().toProto());
+        protoBuilder.setUsesSubspaceKeysCounter(true).setSubspaceKeysCounter(randomCounter);
+        builder = RecordMetaData.newBuilder().setRecords(protoBuilder.build(), true);
+        RecordMetaData metaDataFromProto = builder.build(true);
+        assertNotNull(metaDataFromProto.getIndex("MySimpleRecord$str_value_indexed"));
+        assertNotNull(metaDataFromProto.getIndex("MySimpleRecord$num_value_unique"));
+        assertNotNull(metaDataFromProto.getIndex("MySimpleRecord$num_value_3_indexed"));
+        assertEquals(randomCounter + 1, metaDataFromProto.getIndex("MySimpleRecord$str_value_indexed").getSubspaceKey());
+        assertEquals(randomCounter + 2, metaDataFromProto.getIndex("MySimpleRecord$num_value_unique").getSubspaceKey());
+        assertEquals(randomCounter + 3, metaDataFromProto.getIndex("MySimpleRecord$num_value_3_indexed").getSubspaceKey());
```

Comment:
I see. These indexes are generated from the extension options and *do* get a counter-based value. I think the case I'm worried about is where they are *not* from the extension options (but are added from a previous build).

This is essentially to make sure that a stable upgrade path exists, i.e., that if the user uses an existing meta-data proto definition that their indexes don't move around.

Reply:
Uh, kind of. That test is useful, but one addition that might be needed is to clear out the `subspace_key` field of an index to emulate the behavior of indexes with implicit subspace keys prior to this change to make that that doesn't cause an unexpected subspace key change.

Reply:
see if it's now what you want.

Reply:
Okay, it's almost there. I left a comment on that test explaining why it's not quite there.